### PR TITLE
feat(data-warehouse): add salesforce integration

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- no salesforce integration

## Changes

- adjusts the incremental fields api so that if there are no incremental fields, the set up button still works
- adds salesforce integration flow
- had to modify the onboarding flow slightly because for salesforce, a team needs to give us their subdomain first

TODO:
- decide on how to import custom columns too , right now it's all standard columns. Custom columns require you to provide a limit
- [ ] set client app id and secret in charts

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/temporal/data_imports/pipelines/salesforce/__init__.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            "write_disposition": "merge" if is_incremental else "replace",
+            "endpoint": {
+                "data_selector": "records",
+                "path": "/services/data/v61.0/query",
+                "paginator": SalesforceIncrementalPaginator(schema="Account", subdomain=subdomain),
+                "params": {
+                    "q": {
+                        "type": "incremental",
+                        "cursor_path": "SystemModstamp",
+                        "initial_value": "SELECT FIELDS(STANDARD) FROM Account WHERE SystemModstamp > 2000-01-01T00:00:00.000+0000",
```
Comment (EDsCODE) [inline posthog/temporal/data_imports/pipelines/salesforce/__init__.py:128]: Unclear how to do incremental with this rest query format. I tried to do this but initial_value needs to match the cursor_path type i believe and cursor_path in this case is a datetime.

Oddly enough, the data does return but dlt doesn't write it

Comment (burnash) [reply posthog/temporal/data_imports/pipelines/salesforce/__init__.py:128]: Hey @EDsCODE, 
> initial_value needs to match the cursor_path type i believe and cursor_path in this case is a datetime.

This is true. We're going to allow data conversion for the incremental params, here's the issue: https://github.com/dlt-hub/verified-sources/issues/507

Please take a look to see if the proposed solution fits. Do we miss anything?

Comment (EDsCODE) [reply posthog/temporal/data_imports/pipelines/salesforce/__init__.py:128]: Yep, having a custom function to map the types/parameters would be great.

I think the suggestion in the issue would be enough for my case as we would be receiving a timestamp field and I'd be transforming it into the full query to pass back in

Comment (burnash) [reply posthog/temporal/data_imports/pipelines/salesforce/__init__.py:128]: @EDsCODE the PR with this feature has been merged: https://github.com/dlt-hub/verified-sources/pull/515
Please give it a go. We'll document it soon as well.

File: frontend/src/scenes/data-warehouse/new/NewSourceWizard.tsx [hunk 2]
```diff
@@ -232,9 +232,25 @@ function FirstStep(): JSX.Element {
...
+    let showSourceFields = true
+    if (selectedConnector?.name === 'Salesforce') {
+        if (source.payload.code && source.payload.subdomain) {
+            showPrefix = true
+            showSourceFields = false
+        } else {
+            showPrefix = false
+            showSourceFields = true
+        }
+    }
```
Comment (Gilbert09) [inline frontend/src/scenes/data-warehouse/new/NewSourceWizard.tsx:247]: Not the biggest fan of having this custom logic here - could we add this as a function to the `SOURCE_DETAILS` const in `sourceWizardLogic`? Pass in the `source` object and return whether we hide/show fields

File: frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx [hunk 3]
```diff
@@ -420,6 +421,19 @@ export const SOURCE_DETAILS: Record<ExternalDataSourceType, SourceConfig> = {
...
+        fields: [
+            {
+                name: 'subdomain',
+                label: 'Salesforce subdomain',
+                type: 'text',
+                required: true,
+                placeholder: '',
+            },
+        ],
+        caption: '',
```
Comment (Gilbert09) [inline frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx:435]: Do we want a caption here?

File: frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx [hunk 4]
```diff
@@ -932,16 +985,39 @@ export const sourceWizardLogic = kea<sourceWizardLogicType>([
...
         sourceConnectionDetails: {
             defaults: buildKeaFormDefaultFromSourceDetails(SOURCE_DETAILS),
             errors: (sourceValues) => {
+                if (
+                    values.selectedConnector?.name === 'Salesforce' &&
+                    values.source.payload.code &&
+                    values.source.payload.subdomain
+                ) {
+                    return {}
+                }
```
Comment (Gilbert09) [inline frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx:1008]: I understand why we have this, but 🤮 - we may wanna think about how to make oauth-based sources nicer during onboarding - likely to happen when we onboard more sources

File: posthog/temporal/data_imports/pipelines/salesforce/auth.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            "client_secret": settings.SALESFORCE_CONSUMER_SECRET,
+            "refresh_token": refresh_token,
+        },
+    )
+
+    if res.status_code != 200:
+        err_message = res.json()["error_description"]
+        raise Exception(err_message)
+
+    return res.json()["access_token"]
```
Comment (Gilbert09) [inline posthog/temporal/data_imports/pipelines/salesforce/auth.py:20]: A problem I had with hubspot recently is that the access token is short lived (was 30mins) and so we had to refresh this mid-way through pagination if the user had a lot of data (which one user did). 

We likely wanna call refresh at the start of every sync and not rely on the previously stored access token - what is the length of a salesforce token too? We likely will need to add some custom logic into rest_source to handle refresh tokens too

File: posthog/temporal/data_imports/pipelines/salesforce/helpers.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+"""Salesforce source helpers"""
```
Comment (Gilbert09) [inline posthog/temporal/data_imports/pipelines/salesforce/helpers.py:1]: Is this file still needed now that we're using the rest source?

File: posthog/temporal/data_imports/workflow_activities/import_data.py [hunk 7]
```diff
@@ -216,6 +216,37 @@ async def import_data_activity(inputs: ImportDataActivityInputs):
...
+        from posthog.temporal.data_imports.pipelines.salesforce import salesforce_source
+
+        subdomain = model.pipeline.job_inputs.get("salesforce_subdomain")
+        salesforce_access_token = model.pipeline.job_inputs.get("salesforce_access_token", None)
+        refresh_token = model.pipeline.job_inputs.get("salesforce_refresh_token", None)
+        if not refresh_token:
+            raise ValueError(f"Salesforce refresh token not found for job {model.id}")
+
+        if not salesforce_access_token:
+            salesforce_access_token = salesforce_refresh_access_token(refresh_token)
```
Comment (Gilbert09) [inline posthog/temporal/data_imports/workflow_activities/import_data.py:238]: We likely will need to do this on every run, not only when the access token doesn't exist

File: requirements.in [hunk 8]
```diff
@@ -59,7 +59,7 @@ posthoganalytics==3.5.0
 psycopg2-binary==2.9.7
 PyMySQL==1.1.1
 psycopg[binary]==3.1.18
-pyarrow==15.0.0
+pyarrow==17.0.0
```
Comment (Gilbert09) [inline requirements.in:62]: Any breaking changes with 2 new major versions of pyarrow?

**Review by Gilbert09 — APPROVED** [review]
Rest source stuff mostly looks good - need to figure out refresh tokens and maybe make the UI changes through config

File: posthog/temporal/data_imports/pipelines/salesforce/auth.py (hunk 5 above)
Comment (EDsCODE) [reply posthog/temporal/data_imports/pipelines/salesforce/auth.py:42]: Added custom auth

File: frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx (hunk 4 above)
Comment (EDsCODE) [reply frontend/src/scenes/data-warehouse/new/sourceWizardLogic.tsx:1008]: made it a config also

File: frontend/src/scenes/data-warehouse/new/NewSourceWizard.tsx (hunk 2 above)
Comment (EDsCODE) [reply frontend/src/scenes/data-warehouse/new/NewSourceWizard.tsx:247]: yep good call made it a config

File: requirements.in (hunk 8 above)
Comment (EDsCODE) [reply requirements.in:62]: Not on dw end. DLT with delta requires it now

@tomasfarias note here incase of batch exports impact

Comment (EDsCODE) [reply requirements.in:62]: Removed these updates altogether. Realized they weren't needed. Also, note that the DLT version upgrade causes a test to fail. Errors about s3 configuring missing

**Review by Gilbert09 — APPROVED** [review]
